File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    private ListNode reversedTail; // chatgpt helped me
    private void printListNodeReverse(ListNode head, ListNode reversedHead) {
        // if (head.next != null) {
        //     length = length + 1;
        //     System.out.println(" this is not null ");
        // }
        // if (head.next.next.next == null) {
        //     System.out.println(" this is null ");
        // }
        // return head.next.next.val;
        if (head == null) {
            return;
        }
        printListNodeReverse(head.next, reversedHead);
        System.out.println("val = " + head.val);
        // if (reversedHead.val == 0) {
        //     reversedHead.val = head.val;
        // }
        // System.out.println("val = " + reversedHead.val);

        // chatgpt helped me 
        if (reversedTail == null) {
            reversedHead.val = head.val;
            reversedTail = reversedHead;
        } else {
            reversedTail.next = new ListNode(head.val);
            reversedTail = reversedTail.next;
        }
    }
    public ListNode reverseList(ListNode head) {
        /**
            having hard time with knowing if head is empty
            if head has value and no next
            it means its one [val] so i return it
            if head has no value and no next
            it means its [] so i return it

            other than that i start counting from 1 and
            head.next

            then i get the length from recursion
            lets say its 4
            
            then i attempt to make new ListNode from the end

            for (int i = length - 1; i >= length; i--) {
             yes but here theres no like arr[i]
            }

        */
        // System.out.println("" + head.val);
        if (head == null || head.next == null) { // first successful step alone
            return head;
        }
        int length = 1;
        // System.out.println("" + countLength(head));
        // printListNodeReverse(head); // second successful step alone
        reversedTail = null; // chatgpt helped me
        ListNode reversedHead = new ListNode();
        printListNodeReverse(head, reversedHead); // chatgpt helped me
        return reversedHead;
    }
}

File: Data Structures & Algorithms/longest-common-prefix/submission-0.js
class Solution {
    /**
     * @param {string[]} strs
     * @return {string}
     */
    longestCommonPrefix(strs) {
        const map = {};
        for (let i = 0; i < strs.length; i++) {
            let word = strs[i][0];
            if (map[word]) {
                map[word]++;
            } else {
                map[word] = 1;
            }
            for(let j = 1; j < strs[i].length; j++) {
                word = word + strs[i][j];
                if (map[word]) {
                map[word]++;
                } else {
                map[word] = 1;
                }
            }
        }
        const keys = Object.keys(map);
        let longestCommonPrefix = "";
        for (let i = 0; i < keys.length; i++) {
            if(keys[i].length > longestCommonPrefix.length) {
                if(map[keys[i]] === strs.length) { // chatgpt helped me here after i gave up
                    longestCommonPrefix = keys[i];
                }
            }
        }
        return longestCommonPrefix;
    }
}
